llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,379 @@
+  /// uses. These physical registers should not be used for replacement
+  /// candidates, since this would introduce new WAW dependencies, which this
+  /// pass tries to remove.
+  BitVector getDefinedPhysRegs(const MachineBasicBlock *MBB) const;
+
+  bool replaceReg(const Register Reg, BitVector &BlockedPhysRegs);
+
+  /// Find a free register of the same register class type, but
+  /// exclude the blocked physical registers from the result.
+  /// Otherwise I would introduce new WAW dependencies, that i previously
+  /// removed.
+  MCPhysReg getReplacementPhysReg(const Register Reg,
+                                  const BitVector &BlockedPhysRegs) const;
+
+  bool isWorthRenaming(const Register &Reg, const BitVector &UsedPhysRegs,
+                       const SmallVector<Register, 16> &VRegWithCopies) const;
+
+  /// return the Physical register of the machine operand.
+  MCPhysReg getAssignedPhysReg(const MachineOperand &MO) const;
+
+  /// Identity copies will be removed in a later pass, therefore renaming the
+  /// def here will not remove the copy in a later pass. Thus it increases code
+  /// size without actually being the first define of the physical register.
+  /// $x4 = Copy $x4                                                           \
+  /// This Copy instruction will be removed in a later pass, if the source and
+  /// destination are equal, additionally it is not the first define of $x4 and
+  /// would identify a false WAW hazard, that will be removed by a later pass.
+  bool isIdentityCopy(const MachineInstr &MI) const;
+
+  /// MachineInstructions can write into sub-registers (%0:sub_256_lo and
+  /// %0:sub_256_hi). In the VirtualRegisterMap, they they block the same
+  /// register (i.e. $x0). Which can cause issues, when detecting Write
+  /// After Write (WAW) dependencies. To mitigate this, only use the last
+  /// definition of a virtual register to count the definitions, so that
+  /// writing into subregisters does not already trigger the register
+  /// replacement. In this case $x0 would have already been replaced, even
+  /// though there is no real WAW dependency.
+  ///                                                                          \
+  /// undef %0.sub_256_lo:mxa, _, _= VLDA_2D_dmw_lda_w _,_                     \
+  /// % 0.sub_256_hi:vec512 = VLDA_dmw_lda_w_ag_idx_imm _,_                    \
+  /// %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  ///
+  bool isLastVRegDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                     const MachineBasicBlock &MBB) const;
+
+  /// Block every sub- and super-register of a physical register, so that it is
+  /// removed for future replacement strategies, i.e. block wl4, wh4, y2 if X4
+  /// is used.
+  void addAliasRegs(BitVector &BlockedPhysRegs, const MCPhysReg PhysReg) const;
+};
+
+MCPhysReg
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register Reg = MO.getReg();
```

Comment:
nit: This assumes that it is a register operand. Add it as a precondition to the declaration?

Reply:
Alternatively, we could just pass a Register here.

Reply:
The function now takes a Register as input to make the relationship clear